planner/core/task.go

```diff
@@ -845,22 +858,79 @@ func (p *PhysicalLimit) attach2Task(tasks ...task) task {
 					stats := deriveLimitStats(childProfile, float64(newCount))
 					pushedDownLimit := PhysicalLimit{PartitionBy: newPartitionBy, Count: newCount}.Init(p.SCtx(), stats, p.SelectBlockOffset())
-					pushedDownLimit.SetChildren(partialScan)
+					// when indexPlanFinished is true here, additional limit will only be appended to the table plan.
+					cop = attachPlan2Task(pushedDownLimit, cop).(*copTask)
+					// Don't use clone() so that Limit and its children share the same schema. Otherwise the virtual generated column may not be resolved right.
 					pushedDownLimit.SetSchema(pushedDownLimit.children[0].Schema())
-					limitChildren = append(limitChildren, pushedDownLimit)
+					t = cop.convertToRootTask(p.SCtx())
+					sunk = true
+				} else {
+					// cop.indexPlanFinished = false indicates the table side is a pure table-scan, sink the limit to the index merge index side.
+					newCount := p.Offset + p.Count
+					limitChildren := make([]PhysicalPlan, 0, len(cop.idxMergePartPlans))
+					for _, partialScan := range cop.idxMergePartPlans {
+						childProfile := partialScan.StatsInfo()
+						stats := deriveLimitStats(childProfile, float64(newCount))
+						pushedDownLimit := PhysicalLimit{PartitionBy: newPartitionBy, Count: newCount}.Init(p.SCtx(), stats, p.SelectBlockOffset())
+						pushedDownLimit.SetChildren(partialScan)
+						pushedDownLimit.SetSchema(pushedDownLimit.children[0].Schema())
+						limitChildren = append(limitChildren, pushedDownLimit)
+					}
+					cop.idxMergePartPlans = limitChildren
+					t = cop.convertToRootTask(p.SCtx())
+					sunk = p.sinkIntoIndexMerge(t)
 				}
-				cop.idxMergePartPlans = limitChildren
+			} else {
+				// when there are some root conditions, just sink the limit upon the index merge reader.
+				t = cop.convertToRootTask(p.SCtx())
+				sunk = p.sinkIntoIndexMerge(t)
+			}
+		} else if cop.idxMergeIsIntersection {
+			// In the index merge with intersection case, only the limit can be pushed down to the index merge table side.
+			// Note Difference:
+			// IndexMerge.PushedLimit is applied before table scan fetching, limiting the indexPartialPlan rows returned (it maybe ordered if orderBy items not empty)
+			// TableProbeSide sink limit is applied on the top of table plan, which will quickly shut down the both fetch-back and read-back process.
+			if len(cop.rootTaskConds) == 0 {
+				suspendLimitAboveTablePlan := func() {
+					newCount := p.Offset + p.Count
+					childProfile := cop.plan().StatsInfo()
+					// but "regionNum" is unknown since the copTask can be a double read, so we ignore it now.
+					stats := deriveLimitStats(childProfile, float64(newCount))
+					pushedDownLimit := PhysicalLimit{PartitionBy: newPartitionBy, Count: newCount}.Init(p.SCtx(), stats, p.SelectBlockOffset())
+					// when indexPlanFinished is true here, additional limit will only be appended to the table plan.
+					cop = attachPlan2Task(pushedDownLimit, cop).(*copTask)
```

Comment:
No problem, just seems a little unsafe and complex to call it recursively here... 
Can we just simplify this line to `pushedDownLimit.SetChildren(cop.tablePlan); cop.tablePlan = pushedDownLimit;`?
Removing this recursive function call may make it more comprehensible since we don't need to consider many other statuses or flags here (like `cop.indexPlanFinished` below), we just add a new Limit upon `cop.tablePlan`.

Reply:
definitely sure~